### PR TITLE
Deprecate mutating configurations in buildscript block

### PR DESCRIPTION
**Pull request by jvandort**
The intention of the buildscript block is to contain the 'classpath' configuration and allow dependencies to be declared on the configuration

This commit deprecates using the block for other questionable purposes, by deprecating creation of new configuration in the block.
    
This closes some odd corner cases where a buildscript configuration can resolve other buildscript configurations, a behavior that we never intended to support. There are some questionable use-cases for using the buildscript block to perform adhoc resolutions, for example in settings scripts where there is no other way to obtain a configuration. For this reason, creating detached configurations are still permitted in buildscript blocks.
    
We add a bunch of tests in BuildscriptResolutionIntegrationTest.groovy, which tests corner cases for buildscript configuration resolution.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuilder.java [hunk 1]
```diff
@@ -68,7 +69,9 @@ private List<TaskIdentifier.TaskBasedTaskIdentifier> taskIdentifiersForBuildDepe
         List<TaskIdentifier.TaskBasedTaskIdentifier> tasksToBuild = new ArrayList<>();
         for (Task task : getDependenciesForInternalUse(classpath)) {
             BuildState targetBuild = owningBuildOf(task);
-            assert targetBuild != currentBuild;
+            if (targetBuild == currentBuild) {
```
Comment (jvandort) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuilder.java:72]: One of the tests I wrote triggered the former assertion. It seems the assumption this code made did not hold true. 

It is probably not right to make this an exception without a deprecation cycle, as it seems possible to trigger this condition in user code currently.

**Comment by jvandort** [comment]
@bot-gradle test this without pts

**Comment by jvandort** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuilder.java (hunk 1 above)
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuilder.java:72]: * We should test if the existing behavior runs the task or not
* If the task did not run, we can throw this exception now
* If the task did run, (but how the heck would that happen), then we still throw an exception. This really should not happen.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java [hunk 2]
```diff
@@ -158,6 +159,12 @@ private void defineConfiguration() {
         }
         if (classpathConfiguration == null) {
             classpathConfiguration = configContainer.migratingUnlocked(CLASSPATH_CONFIGURATION, ConfigurationRolesForMigration.LEGACY_TO_RESOLVABLE_DEPENDENCY_SCOPE);
+            configContainer.beforeCollectionChanges(methodName ->
```
Comment (tresat) [inline subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java:162]: 🤔 If we have the methodName available here, why not work that into the error message itself?  This would help people who encounter this error quickly figure out what's going on with it.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,832 @@
...
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.test.fixtures.dsl.GradleDsl
+
+/**
+ * Tests edge cases of buildscript configuration resolution.
+ *
+ * <p>Tests should cover cases of initscript, settings, standalone, and project buildscript configurations.</p>
+ */
+class BuildscriptResolutionIntegrationTest extends AbstractIntegrationSpec {
```
Comment (tresat) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy:27]: 🤔 I'd consider breaking this class into 4, one for each type os script.

**Review by tresat — CHANGES_REQUESTED** [review]
I think you need to cover the case of non-create mutations also now failing in the upgrade guide.

Other than that, this change looks good to me.  The rest of my comments are minor.

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 4]
```diff
@@ -85,6 +85,56 @@ We decided to upgrade the minimum version of the Java runtime for a number of re
...
+    }
+}
+
+val files = buildscript.configurations["myConfig"].files
+----
+======
+=====
+
+This pattern is sometimes used to resolve dependencies in Settings, where there is no other way to obtain a Configuration.
+Resolving dependencies in this context is not recommended, however using a detached configuration will still work for the time being.
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:119]: I wouldn't necessarily _recommend_ this alternative. If you have to do it this is the only thing you can do really. 

Is there another wording you'd prefer that also reflects that?

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RootComponentResolutionIntegrationTest.groovy [hunk 5]
```diff
@@ -326,4 +215,36 @@ class RootComponentResolutionIntegrationTest extends AbstractIntegrationSpec {
         expect:
         succeeds("resolve")
     }
+
+    // This is not necessarily desired behavior, or important behavior at all.
+    // The detached configuration is _not_ the project. It should not claim to be the project.
+    // Ideally, this configuration would have an unspecified identity, similar to init, settings, and standalone scripts.
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/RootComponentResolutionIntegrationTest.groovy:221]: One day

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java [hunk 6]
```diff
@@ -400,4 +403,9 @@ protected I createDomainObject() {
             return factory.apply(getName());
         }
     }
+
+    @Override
+    public String getDisplayName() {
+        return Describables.of("configuration container for", owner).getDisplayName();
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java:409]: Agreed, can simplify this

File: subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultDomainObjectCollectionTest.groovy [hunk 7]
```diff
@@ -426,6 +426,137 @@ class DefaultDomainObjectCollectionTest extends AbstractDomainObjectCollectionSp
         !container.remove("a")
     }
 
+    def callsVetoActionBeforeObjectIsAdded() {
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/DefaultDomainObjectCollectionTest.groovy:429]: I think the prior functionality intended these methods to always "veto" mutations by throwing exceptions when they did not want the mutation to occur. That is where that wording came from. I can adjust these method names though

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java [hunk 8]
```diff
@@ -67,13 +67,13 @@ public T maybeCreate(String name) {
...
     public T create(String name, Closure configureClosure) {
-        assertMutable("create(String, Closure)");
+        assertCanMutate("create(String, Closure)");
         return create(name, ConfigureUtil.configureUsing(configureClosure));
     }
 
     @Override
     public T create(String name, Action<? super T> configureAction) throws InvalidUserDataException {
-        assertMutable("create(String, Action)");
+        assertCanMutate("create(String, Action)");
```
[the remark this answers is not in the document]
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java:76]: That would mean assertCanAdd would need to accept both the method name and the name of thing we are trying to add.

Really I think assertCanAdd is a bad name. I would expect that if I call `assertCanAdd` that means I can add something. To satisfy that assertion, as you mention, that should verify 
1. An object with that name is not present
2. The container is mutable

I think I would prefer to keep these operations separate, particularly since there are instances where `assertCanAdd` are currently called without an `assertCanMutate` (presumably since we've already checked that we can mutate elsewhere)

I would consider renaming this to something more accurate, like `assumeElementNotPresent` or similar

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java (hunk 2 above)
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultScriptHandler.java:162]: I probably should be able to integrate this into the message

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy (hunk 3 above)
Comment (jvandort) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy:27]: I think ATM I would prefer to keep these in one place. If somebody adds tests that verify some behavior for one kind script, I'd like them to also verify them for the other types of scripts as well. 

Having them all in the same place means the tests for all 4 types of scripts are discoverable, but having them in separate places could make it easy to write a test for one script but not the others. 

I would consider an abstract class based solution, but at the moment I dont want to develop that infrastructure

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 9]
```diff
@@ -85,6 +85,56 @@ We decided to upgrade the minimum version of the Java runtime for a number of re
...
+
+The buildscript configurations block is only intended to be used to declare dependencies for the buildscript classpath.
+
+Consider the following script that creates a new buildscript configuration in a Settings script and resolves it:
+
+=====
+[.multi-language-sample]
+======
+.settings.gradle.kts
+[source,kotlin]
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:102]: Will update this to single-language

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (tresat) [reply platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:119]: Resolving dependencies in this context is not recommended.  Using a detached configuration is a possible alternative.

File: subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/AbstractNamedDomainObjectContainer.java:76]: Makes sense.

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy (hunk 3 above)
Comment (tresat) [reply platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/BuildscriptResolutionIntegrationTest.groovy:27]: An abstract class solution seems like a good fit, but yeah, not a big concern.

File: subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuilder.java (hunk 1 above)
Comment (jvandort) [reply subprojects/core/src/main/java/org/gradle/api/internal/initialization/DefaultBuildLogicBuilder.java:72]: This fails with a different exception if we remove the assertion and new exception. 

Leaving this as-is

**Comment by jvandort** [comment]
@bot-gradoe test prf bd

**Comment by jvandort** [comment]
@bot-gradle test prf bd

**Review by tresat — APPROVED** [review]

**Comment by jvandort** [comment]
@bot-gradle test prf

**Comment by jvandort** [comment]
@bot-gradle test ast
